### PR TITLE
Incorporated trees and purifiers module.

### PR DESCRIPTION
**Pull request by madDipanjan**

### CODE REVIEW

File: tools/IBM/writeIBMFiles.m
Comment (samoliverowens) [inline tools/IBM/writeIBMFiles.m:1]: What does 'Range' do here?

File: tools/IBM/writeIBMFiles.m [hunk 1]
```diff
@@ -853,10 +853,10 @@
 ylim([0 ysize])
 zlim([0 zsize])
 
-%scatter3(X_u(solid_u), Y_u(solid_u), Z_u(solid_u), 10,[0,0,1],'filled')
-%scatter3(X_v(solid_v), Y_v(solid_v), Z_v(solid_v), 10,[0,0,1],'filled')
-%scatter3(X_w(solid_w), Y_w(solid_w), Z_w(solid_w), 10,[0,0,1],'filled')
-%scatter3(X_c(solid_c), Y_c(solid_c), Z_c(solid_c), 10,[0,0,1],'filled')
+scatter3(X_u(solid_u), Y_u(solid_u), Z_u(solid_u), 10,[0,0,1],'filled')
```
Comment (samoliverowens) [inline tools/IBM/writeIBMFiles.m:856]: I would prefer if the preprocessing didn't try to plot anything by default... We could potentially put a switch around this.

File: tools/gather_outputs.sh [hunk 2]
```diff
@@ -50,13 +50,15 @@ for file in *dump.*.000.${expnr}.nc ; do
         dumps=${file%.000.${expnr}.nc}
 
         if [ ${dumps:0:9} == "fielddump" ]; then
-            ymparam="v,tau_y,ym"
+            #ymparam="v,tau_y,ym"
```
Comment (samoliverowens) [inline tools/gather_outputs.sh:53]: Note if 'tau_x' is specified as a fielddump variable, this line needs to be used, but now that we have stress output on facets this is probably needed less, so let's keep it this way.

File: tools/nco_concatenate_field_x.sh
Comment (samoliverowens) [inline tools/nco_concatenate_field_x.sh:1]: Why is this changed needed? It seems to me that in `gather_output.sh`, we call `nco_concatenate_field_y.sh` first, which generates files of the form `dump.???.nc`.

File: tools/preprocessing.m [hunk 3]
```diff
@@ -140,20 +140,30 @@ function chcpath(obj, newpath)
         function set_defaults(obj)
             %% &RUN
             preprocessing.addvar(obj, 'ltrees', 0) % switch for trees (not implemented)
```
Comment (samoliverowens) [inline tools/preprocessing.m:142]: Remove this 'not implemented' :)

File: tools/preprocessing.m [hunk 4]
```diff
@@ -211,6 +221,7 @@ function set_defaults(obj)
             %% &WALLS
             preprocessing.addvar(obj, 'iwallmom', 3)
             preprocessing.addvar(obj, 'iwalltemp', 1)
+            preprocessing.addvar(obj, 'lbottom', 0)
```
Comment (samoliverowens) [inline tools/preprocessing.m:224]: Don't think `lbottom` is actually needed in preprocessing, but fine to have it anyway.

File: tools/write_inputs.m [hunk 5]
```diff
@@ -20,10 +20,12 @@
...
 
-expnr = '136';
+expnr = '901';
 %
-DA_EXPDIR = getenv('DA_EXPDIR');
-DA_TOOLSDIR = getenv('DA_TOOLSDIR');
+% DA_EXPDIR = getenv('DA_EXPDIR');
+% DA_TOOLSDIR = getenv('DA_TOOLSDIR');
+DA_EXPDIR = 'D:/Postdoc1/simulation/treeu2/experiments';
+DA_TOOLSDIR = 'D:/Postdoc1/simulation/treeu2/u-dales/tools';
```
Comment (samoliverowens) [inline tools/write_inputs.m:28]: These paths need to be reset.

File: tools/write_inputs.m [hunk 6]
```diff
@@ -75,7 +83,7 @@
 
     calculate_facet_sections_uvw = r.iwallmom > 1;
     calculate_facet_sections_c = r.ltempeq || r.lmoist;
-    lwindows = false;
+    lwindows = true;
```
Comment (samoliverowens) [inline tools/write_inputs.m:86]: Reset to 'false'.

File: tools/write_inputs.m [hunk 7]
```diff
@@ -175,7 +183,7 @@
         resolution   = r.psc_res;
         xazimuth     = r.xazimuth;
         ltimedepsw   = r.ltimedepsw;
-        ldirectShortwaveFortran = 0;
+        ldirectShortwaveFortran = 1;
```
Comment (samoliverowens) [inline tools/write_inputs.m:186]: Reset to '0' for now.

**Review by samoliverowens — CHANGES_REQUESTED** [review]

File: tools/IBM/writeIBMFiles.m
Comment (madDipanjan) [reply tools/IBM/writeIBMFiles.m:1]: 'Range', 2 in the readmatrix command tells it to start reading from the second line onward, skipping the first line. In our files, the first line has headers. When there is no entry on the txt file, e.g. for the case of a flat ground stl, readmatrix reads the header and returns nan instead of returning a empty matrix, if Range is not in use. So, added the Range keyword.

File: tools/nco_concatenate_field_x.sh
Comment (madDipanjan) [reply tools/nco_concatenate_field_x.sh:1]: Because the file format currently is in this format considering 2d decomposition, treedump.***.***.001.nc for example. See at lines 50 and 59 where it was correctly done earlier but at line 38 we missed it earlier. I encountered this while merging treedump and corrected it. We didn't see this for tdump of fielddump because these cases do not go inside the if clause of  of line 36.

File: tools/write_inputs.m (hunk 7 above)
Comment (madDipanjan) [reply tools/write_inputs.m:186]: Reset done.

File: tools/write_inputs.m (hunk 6 above)
Comment (madDipanjan) [reply tools/write_inputs.m:86]: Reset done.

File: tools/write_inputs.m (hunk 5 above)
Comment (madDipanjan) [reply tools/write_inputs.m:28]: Corrected. Sorry my mistake :)

File: tools/preprocessing.m (hunk 3 above)
Comment (madDipanjan) [reply tools/preprocessing.m:142]: Removed

File: tools/preprocessing.m (hunk 4 above)
Comment (madDipanjan) [reply tools/preprocessing.m:224]: Actually I was using it for some tree preprocessing. But later removed that part. Let it be as it is as I might again work on the tree preprocessing. This will be mainly be needed for plotting purpose. I am thinking of having a section at the end of write_inputs for all the check plots. Will see later.

File: tools/gather_outputs.sh (hunk 2 above)
Comment (madDipanjan) [reply tools/gather_outputs.sh]: Modified it appropriately. Kept in this way

ymparam="v,tau_y,ym"
#ymparam="v,ym"

File: tools/IBM/writeIBMFiles.m (hunk 1 above)
Comment (madDipanjan) [reply tools/IBM/writeIBMFiles.m]: Commented these lines.

**Review by samoliverowens — APPROVED** [review]

**Review by samoliverowens — APPROVED** [review]
